File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    answer = -1
    can_make = []
    
    for i in range (1,9):
        case = set()
        case.add(int(str(N) * i))
        for j in range(i-1):
            for num1 in can_make[j]:
                for num2 in can_make[i-j-2]:
                    case.add(num1 - num2)
                    case.add(num1 + num2)
                    case.add(num1 * num2)
                    if num2 != 0:
                        case.add(num1 // num2)

        if number in case:
            answer = i
            break
            
        can_make.append(case)
        
    return answer

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        int answer = -1;
        
        ArrayList<Set<Integer>> canMake = new ArrayList<>();
        for(int i = 0; i <= 8; i++) {
            canMake.add(new HashSet<>());
        }
        
        for(int i = 1; i <= 8; i++) {
            HashSet<Integer> eachCase = new HashSet<>();
            eachCase.add(Integer.parseInt(Integer.toString(N).repeat(i)));
            for(int j = 1; j < i; j++) {
                for(Integer num1 : canMake.get(j)) {
                    for(Integer num2 : canMake.get(i-j)) {
                        eachCase.add(num1 + num2);
                        eachCase.add(num1 - num2);
                        eachCase.add(num1 * num2);
                        if(num2 != 0) eachCase.add(num1 / num2);
                    }
                }
            }

            if(eachCase.contains(number)) {
                answer = i;
                break;
            }
            
            canMake.set(i, eachCase);
        }
        
        return answer;
    }
}